### PR TITLE
Remove usage of 'keyword_list'

### PR DESCRIPTION
**Pull request by ghost**
We're using `keyword_list` to generate code for some of the animation properties. This could be achieved with `single_keyword` (passing `vector=True`).

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach build-geckolib` does not report any errors

- [x] These changes do not require tests because it's a cleanup

r? @emilio

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13917)

### CODE REVIEW

File: components/style/properties/helpers.mako.rs [hunk 1]
```diff
@@ -75,6 +75,10 @@
             }
             pub mod computed_value {
                 pub use super::single_value::computed_value as single_value;
+                % if requires_single:
+                pub use super::single_value::computed_value::T as SingleComputedValue;
```
Comment (Manishearth) [inline components/style/properties/helpers.mako.rs]: I don't think we need `requires_single`. Refactor the code that needs `SingleFoo` to use `single_value::computed_value::Foo`. Alternatively, unconditionally add the reexport.

File: components/style/properties/helpers.mako.rs [hunk 2]
```diff
@@ -145,6 +151,19 @@
                     }).map(SpecifiedValue)
                 % endif
             }
+
+            % if requires_single:
```
Comment (Manishearth) [inline components/style/properties/helpers.mako.rs]: We should get rid of this and refactor code using these functions to use the `single_value` functions instead

**Comment by Wafflespeanut** [comment]
I really thought I could escape from refactoring a lot of code :laughing:

File: components/style/properties/helpers.mako.rs [hunk 3]
```diff
@@ -75,6 +74,7 @@
             }
             pub mod computed_value {
                 pub use super::single_value::computed_value as single_value;
+                pub use super::single_value::computed_value::T as SingleComputedValue;
```
Comment (emilio) [inline components/style/properties/helpers.mako.rs]: nit: `pub use single_value::T as SingleComputedValue` is shorter.

File: components/style/properties/helpers.mako.rs [hunk 4]
```diff
@@ -51,9 +51,8 @@
     We assume that the default/initial value is an empty vector for these.
     `initial_value` need not be defined for these.
 </%doc>
-<%def name="vector_longhand(name, gecko_only=False, allow_empty=False, **kwargs)">
+<%def name="vector_longhand(name, allow_empty=False, **kwargs)">
     <%call expr="longhand(name, **kwargs)">
-        % if product == "gecko" or not gecko_only:
```
Comment (emilio) [inline components/style/properties/helpers.mako.rs]: Deindent?

**Review by emilio — COMMENTED** [review]
Just one question, that I think @SimonSapin should be able to answer.

So the idea of not having `T::get_initial_value`, as far as I understand, is that you might want to properties with the same type, but different initial values for each of them, right?

If that's the case, that change probably should be reverted. Otherwise the simplification looks pretty neat :)

**Comment by Manishearth** [comment]
> If that's the case, that change probably should be reverted.

**Comment by SimonSapin** [comment]
> Just one question, that I think @SimonSapin should be able to answer.

I have no idea what is the context or motivation for this PR. (By the way, it would be nicer to include that in the PR message.) But yes, in CSS specs initial values are define per individual property, even if some properties may share some syntax / grammar.

**Comment by Manishearth** [comment]
> I have no idea what is the context or motivation for this PR. 

We have both `helpers.keyword(vector=True)` and `helpers.keyword_list()` which do the same thing but slightly differently.

I guess we shouldn't move get_initial to a method, then.

**Comment by Wafflespeanut** [comment]
> So the idea of not having `T::get_initial_value`, as far as I understand, is that you might want to properties with the same type, but different initial values for each of them, right?

No, it was because initially, I was unable to have the `get_initial_value` function in each module. Now, I've done a workaround. `T::get_initial_value` is no longer a method :)

**Comment by Manishearth** [comment]
> No, it was because initially, I was unable to have the get_initial_value function in each module. Now, I've done a workaround. T::get_initial_value is no longer a method :)

I still don't get why this is necessary. The `vector=True` helper creates a single_value mod, just put the second function inside that?

**Comment by Wafflespeanut** [comment]
Updated the PR message :)

**Comment by Manishearth** [comment]
You should rewrite callers to use `property::single_value::computed_value::get_initial`. Don't add a conflicting alias.

**Comment by Manishearth** [comment]
Or, `property::computed_value::single_value::get_initial`. Both work.

File: components/style/properties/longhand/box.mako.rs [hunk 5]
```diff
@@ -738,6 +730,20 @@ ${helpers.single_keyword("overflow-x", "visible hidden scroll auto",
             Infinite,
         }
 
+        impl Parse for AnimationIterationCount {
+            fn parse(input: &mut ::cssparser::Parser) -> Result<Self, ()> {
+                if input.try(|input| input.expect_ident_matching("infinite")).is_ok() {
+                    Ok(AnimationIterationCount::Infinite)
```
Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: nit: return early and deindent the next block if you wish.

File: components/style/properties/longhand/box.mako.rs [hunk 6]
```diff
@@ -640,12 +632,24 @@ ${helpers.single_keyword("overflow-x", "visible hidden scroll auto",
...
     pub mod computed_value {
-        use std::fmt;
         use Atom;
+        use parser::Parse;
+        use std::fmt;
         use style_traits::ToCss;
 
         pub use Atom as SingleComputedValue;
 
+        impl Parse for Atom {
```
Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: Yes, writing a newtype for this is adequate I think

Comment (Wafflespeanut) [inline components/style/properties/longhand/box.mako.rs]: Okay, I've gone for a new type `AnimationName(pub Atom)`

File: components/style/properties/helpers.mako.rs (hunk 3 above)
Comment (Wafflespeanut) [inline components/style/properties/helpers.mako.rs]: Yeah, I tried, it doesn't seem to work?

    error[E0432]: unresolved import `single_value::T`. Maybe a missing `extern crate single_value`?
       --> /media/Ubuntu/servo/target/geckolib/debug/build/style-8deb288d0c4a7300/out/properties.rs:315:25
        |
    315 |                 pub use single_value::T as SingleComputedValue;
        |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

**Comment by Wafflespeanut** [comment]
Updated the PR.

File: components/style/properties/longhand/box.mako.rs [hunk 7]
```diff
@@ -640,15 +632,49 @@ ${helpers.single_keyword("overflow-x", "visible hidden scroll auto",
...
+        #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+        pub struct AnimationName(pub Atom);
+
+        impl fmt::Display for AnimationName {
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+                self.0.fmt(f)
+            }
+        }
+
+        impl Deref for AnimationName {
```
Comment (Manishearth) [inline components/style/properties/longhand/box.mako.rs]: In general try to use deref for things that involve actual dereferencing. Just refactor everything else to use `.0` here

Comment (Wafflespeanut) [inline components/style/properties/longhand/box.mako.rs]: Removed the impls.

**Pull request merged**